server/events/project_command_context_builder.go

```diff
@@ -298,14 +322,30 @@ func getTfVersion(ctx *command.Context, absProjDir string) *version.Version {
 	requiredVersionSetting := module.RequiredCore[0]
 	ctx.Log.Debug("found required_version setting of %q", requiredVersionSetting)
 
-	tflist, _ := lib.GetTFList(mirrorURL, true)
-	constrains, _ := semver.NewConstraint(requiredVersionSetting)
-	versions := make([]*semver.Version, len(tflist))
+	tfVersions, err := terraformClient.ListAvailableVersions(ctx.Log)
+	if err != nil {
+		ctx.Log.Err("Unable to list Terraform versions")
+	}
+
+	if len(tfVersions) == 0 {
+		// Fall back to an exact required version string
+		// We allow `= x.y.z`, `=x.y.z` or `x.y.z` where `x`, `y` and `z` are integers.
+		re := regexp.MustCompile(`^=?\s*([0-9.]+)\s*$`)
+		matched := re.FindStringSubmatch(requiredVersionSetting)
+		if len(matched) == 0 {
+			ctx.Log.Debug("Did not specify exact version in terraform configuration, found %q", requiredVersionSetting)
+			return nil
+		}
+		tfVersions = []string{matched[1]}
+	}
```

Comment:
If return nil from this method, atlantis try to use default version.

https://github.com/runatlantis/atlantis/blob/f680e4b4a4868a0977cadce16fb8003ae923c5a6/server/core/runtime/plan_step_runner.go#L35-L39

So I'm not sure this logic is required? If users always want to use default version, this logic may cause `terraform$version command not found`.

Reply:
Normally, we should only end up here if downloads are disabled, in which case we don't necessarily want to fall back to the default version. If people pre-bake their `terraform` binaries into their Atlantis image, then this still allows Atlantis to pick the appropriate version.

However, I missed that the `ListAvailableVersions` call doesn't propagate its error (it just logs it and soldiers on), so we could end up here if that call somehow fails in a graceful manner (though, at present, it tends to fail by calling `os.Exit` 🙃 ). So, we could either:
* Propagate that error, or
* Return `nil` if there's an error, in which case it will try to continue with the default version.

Reply:
[Atlantis provided docker images contains default version](https://github.com/runatlantis/atlantis/blob/9a38e354936fbf93282e4e934fa3439fc3bfd102/Dockerfile#L22-L45), so almost cases we got default version, I think. But if someone use custom image, it's true.

> err

terraform-switcher os.Exit is a little bit trouble for me. To be honnest, if we use terraform-switcher, it is better to check we can access them before call terraform-switcher lib.

```diff
+      // terraform-switcher call os.Exit(1) if we cannot get data such as airgapped environment.
+      // before call it, we just check it's possible to call them.
+       resp, err := http.Get(url)
+       if err != nil || resp.StatusCode != 200 {
+               return nil, fmt.Errorf("...")
+       }
        versions, err := lib.GetTFList(url, true)
```

After that time we can propagate err. Even if we disable download, ListAvailables returns `nil, nil`, so no problem.

Reply:
I think this approach, but it's generating a gosec error 🤦 

I'm not that familiar with the golang ecosystem, so I'm sure if there's an easy way to appease that check or not.